### PR TITLE
feat(curriculum): add functional programming lab to new cert

### PR DESCRIPTION
**Pull request by Dario-DC**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

**Comment by moT01** [comment]
A PR has been merged that renamed the certification from "Front End" to "Full Stack". This PR may need an update and a few changes made: [see comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/56694#issuecomment-2444473378)

**Comment by moT01** [comment]
A PR has been merged that moves the block order from the `meta.json` file to the `full-stack.json` file.](https://github.com/freeCodeCamp/freeCodeCamp/pull/56795) This PR will need an update, and the order will need to be removed from the `meta.json` file and moved to the new file.

File: curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+1. You should have a function named `isOrdered` that takes two integers and returns a boolean indicating if the first integer is less than the second.
+1. You should have a function named `swapElements` that takes an array of integers and a numeric index.
+1. The `swapElements` function should modify the array in place by swapping the element at the passed index and the following element if the first element is greater than the second.
+1. You should have a function named `highligthCurrentEls` that takes an HTML element and a numeric index.
+1. The `highligthCurrentEls` function should set the border of the given element's descendants placed at the given index and at the next one to `2px dashed red`.
+1. When you click `#generate-btn` you should fill `#starting-array-container` with five `span` elements, each with a random number as its text.
+1. After you click `#sort-btn`, `#array-container` should contain a `div` element for each of the steps required by the Bubble Sort algorithm to sort the starting array, including the `div` representing the starting array and a `div` representing the sorted array.
+1. Each `div` should contain five `span`, each with a number as its text.
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md:31]:
```suggestion
1. Each `div` should contain five `span` elements, representing the array in its current state of being sorted.
```

**Review by moT01 — CHANGES_REQUESTED** [review]
This is looking good @Dario-DC 👍 It was quite fun and challenging. I was a little confused with some of the things I needed to do, but was able to figure it out. I left a number of typo and grammar suggestions below to try and make some things a little more clear.

I wasn't actually able to get all the tests to pass - even though it seemed like I completed the user stories and tests as described - but maybe I was missing something...

<details><summary>Here's the code I used</summary>

```js
// Elements
const generateEl = document.getElementById('generate-btn');
const startContainerEl = document.getElementById('starting-array-container');
const sortEl = document.getElementById('sort-btn');
const containerEl = document.getElementById('array-container');

// Helper functions
const generateElement = () => Math.floor(Math.random() * 100) + 1;
const generateArray = () => Array.from({ length: 5 }, generateElement);
const generateContainer = () => document.createElement("div");
const fillArrContainer = (arr, el) => arr.forEach(a => el.append(Object.assign(document.createElement('span'), { innerText: a })));
const isOrdered = (a, b) => a < b;
const swapElements = (arr, i) => arr[i] > arr[i + 1] && ([arr[i], arr[i + 1]] = [arr[i + 1], arr[i]]);
const highligthCurrentEls = (el, i) => {
  [el.children[i], el.children[i + 1]].forEach(child => child.style.border = '2px dashed red');
};

let currentArr = [];

// Listeners
generateEl.addEventListener('click', () => {
  while (containerEl.lastChild && containerEl.children.length > 1) {
    containerEl.removeChild(containerEl.lastChild);
  }
  startContainerEl.innerHTML = '';
  currentArr = generateArray();
  fillArrContainer(currentArr, startContainerEl);
});

sortEl.addEventListener('click', () => {
  let isStartingDiv = true;
  let didSwap;

  do {
    didSwap = false;

    for (let i = 0; i < currentArr.length - 1; i++) {
      if (isStartingDiv) {
        highligthCurrentEls(startContainerEl, i);
        isStartingDiv = false;
      } else {
        const newDivEl = generateContainer();
        fillArrContainer(currentArr, newDivEl)
        highligthCurrentEls(newDivEl, i)
        containerEl.append(newDivEl)
      }

      if (!isOrdered(currentArr[i], currentArr[i+1])) {
        didSwap = true;
        swapElements(currentArr, i);
      }

    }
  } while (didSwap)  

  const newDivEl = generateContainer();
  fillArrContainer(currentArr, newDivEl)
  containerEl.append(newDivEl)
});
```

</details>

<details><summary>And here's the tests that didn't pass</summary>

Screenshot 2024-11-14 at 9 47 21 PM

</details>

Finally, perhaps some of the CSS could be adjusted a little. Depending on the number of steps it takes to sort and the size of the window, some of the elements were getting cut off...

<details><summary>image</summary>

Screenshot 2024-11-14 at 10 14 33 PM

^ the divs with the arrays, and the buttons to generate and sort.

</details>

**Comment by Dario-DC** [comment]
Hey @moT01 after modifying the id and fixing the typo in the function name I used your code to check those failing tests.
Here's the only one that fails for me:
image
Not sure why it fails and why the others that failed for you are passing :thinking:

**Comment by moT01** [comment]
I took another small look at this. I found a couple of my tests weren't passing because of my browser zoom level. The `2px dashed red` border was calculated at `1.65px dashed red` or something - making some tests fail. That last test you mentioned, I couldn't quite figure out. It has to do with me calling the `swapElements` function only `if (!isOrdered(currentArr[i], currentArr[i+1]))`, but that array never gets generated because the generate button isn't clicked in the test. I added that, and it gave my an infinite loop. Maybe we should just remove that test for now? I feel like others might run into similar issues.

**Comment by Dario-DC** [comment]
Hey @moT01 , I was able to make that failing test work with your code but I was still getting "potential infinite loop on line ..." so I removed it as you suggested.

I slightly modified the CSS. Not sure if that is enough.

Should I change the tests that check that the border is `2px dashed red`?

**Comment by ilenia-magoni** [comment]
> Not strictly related to this: why don't we have the reset your code button anymore? If someone deletes something from the seed code and saves the changes it will be almost impossible to pass the lab then.

this deserves an issue of its own, imho
With many of the projects having a lot of boilerplate code that is needed to make things work, having a `Reset Code` button seems important

File: curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+    sortBtn.dispatchEvent(new Event("click"));
+    const container = document.querySelector("#array-container");
+    assert.lengthOf(container.children, 13)
+    Array.from(container.children).forEach(el => {assert.equal(el.tagName, "DIV")})
+} finally {
+    randomMocker.restore();
+}
+```
+
+After you click `#sort-btn`, each `div` within `#array-container` should contain five `span`, each with number as its text, and arranged to represent the steps required by Bubble Sort algorithm to sort the starting array.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md:250]:
```suggestion
After you click `#sort-btn`, each `div` within `#array-container` should contain five `span`, each with a number as its text, and arranged to represent the steps required by Bubble Sort algorithm to sort the starting array.
```

**Review by moT01 — APPROVED** [review]
LGMT 🎉 

The only thing we could maybe add is a mention that the browser needs to be at 100% zoom to pass - at least for firefox.

I also see some strange changes in the intro.json file - I would be fine merging this and we can audit that file after.

File: client/i18n/locales/english/intro.json
Comment (ilenia-magoni) [inline client/i18n/locales/english/intro.json:1]: what happened to this file?

**Review by majestic-owl448 — COMMENTED** [review]
That is the only comment I can make, I did not manage to complete this lab nor to understand why my solution doesn't work

File: curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+    sortBtn.dispatchEvent(new Event("click"));
+    const container = document.querySelector("#array-container");
+    assert.lengthOf(container.children, 13)
+    Array.from(container.children).forEach(el => {assert.equal(el.tagName, "DIV")})
+} finally {
+    randomMocker.restore();
+}
+```
+
+After you click `#sort-btn`, each `div` within `#array-container` should contain five `span`, each with a number as its text, and arranged to represent the steps required by Bubble Sort algorithm to sort the starting array.
```
Comment (zairahira) [inline curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md:250]: This test is passing on initial seed code.

**Review by zairahira — COMMENTED** [review]
I wasn't able to pass tests 15 and 20 with the provided solution.

File: curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+    finalArr
+]
+const genBtn = document.querySelector("#generate-btn");
+const sortBtn = document.querySelector("#sort-btn");
+const randomMocker = new __helpers.RandomMocker();
+randomMocker.mock();
+try {    
+    genBtn.dispatchEvent(new Event("click"));
+    sortBtn.dispatchEvent(new Event("click"));
+    const container = document.querySelector("#array-container");
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md:276]:
```suggestion
    const container = document.querySelector("#array-container");
    assert.isNotEmpty(container.children);
```
maybe

**Comment by Dario-DC** [comment]
> I wasn't able to pass tests 15 and 20 with the provided solution.

Could it be related to this?  :arrow_down: @zairahira 
> The only thing we could maybe add is a mention that the browser needs to be at 100% zoom to pass - at least for firefox.

File: client/i18n/locales/english/intro.json
Comment (Dario-DC) [reply client/i18n/locales/english/intro.json:1]: I removed duplicate placeholders and sorted the blocks

**Comment by Dario-DC** [comment]
Hey @zairahira I removed the 2px requirement. Let me know if that worked for you.

**Comment by zairahira** [comment]
@Dario-DC,
The tests are now passing on Firefox and Chrome both 🎉

**Comment by Dario-DC** [comment]
I resolved the conflicts. Despite the changes in the `intro.json` the order should be correct @zairahira

File: curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+id: 6716249b5405164036fd0b0d
+title: Build a Sorting Visualizer
+challengeType: 25
+dashedName: build-a-sorting-visualizer
+demoType: onClick
+---
+
+# --description--
+
+The Bubble Sort algorithm sorts a sequence of integers by comparing couples of adjacent elements starting from the beginning of the sequence. If the first element is greater then the second one, it swaps them. Then, it proceeds with the following couple. When the sequence ends, it starts a new cycle from the beginning of the sequence, and repeats the process until the elements are sorted.
```
Comment (Sembauke) [inline curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md:11]:
```suggestion
The Bubble Sort algorithm sorts a sequence of integers by comparing couples of adjacent elements starting from the beginning of the sequence. If the first element is greater than the second one, it swaps them. Then, it proceeds with the following couple. When the sequence ends, it starts a new cycle from the beginning of the sequence, and repeats the process until the elements are sorted.
```

**Review by Sembauke — APPROVED** [review]

**Pull request merged**